hibernate-core/src/main/java/org/hibernate/internal/util/config/ConfigurationHelper.java

```diff
@@ -462,17 +462,16 @@ public static String resolvePlaceHolder(String property) {
 				// peek ahead
 				if ( chars[pos+1] == '{' ) {
 					// we have a placeholder, spin forward till we find the end
-					String systemPropertyName = "";
 					int x = pos + 2;
-					for (  ; x < chars.length && chars[x] != '}'; x++ ) {
-						systemPropertyName += chars[x];
+					while ( x < chars.length && chars[x] != '}' ) {
 						// if we reach the end of the string w/o finding the
 						// matching end, that is an exception
 						if ( x == chars.length - 1 ) {
 							throw new IllegalArgumentException( "unmatched placeholder start [" + property + "]" );
 						}
+						x++;
 					}
-					String systemProperty = extractFromSystem( systemPropertyName );
+					String systemProperty = extractFromSystem( property.substring( pos + 2, x ) );
```

Comment:
Not 100% sure about the correctness, but the original `systemPropertyName += chars[x];` might be too inefficient.